zerver/lib/bugdown/__init__.py

```diff
@@ -1040,11 +1048,29 @@ def find_proper_insertion_index(self, grandparent: Element, parent: Element,
     def is_absolute_url(self, url: str) -> bool:
         return bool(urllib.parse.urlparse(url).netloc)
 
+    def maybe_append_attachment_url(self, url: str) -> None:
+        attachment_regex = get_compiled_attachment_regex()
+        m = attachment_regex.match(url)
+        if m:
+            urls = self.markdown.zulip_message.potential_attachment_urls
+            urls.append(m.group('path'))
+            self.markdown.zulip_message.potential_attachment_urls = urls
+
     def run(self, root: Element) -> None:
         # Get all URLs from the blob
         found_urls = walk_tree_with_family(root, self.get_url_data)
         # Collect unique URLs which are not quoted
         unique_urls = {found_url.result[0] for found_url in found_urls if not found_url.family.in_blockquote}
+
+        # Set has_link and similar flags whenever a message is processed by bugdown
+        if self.markdown.zulip_message:
+            self.markdown.zulip_message.has_link = len(found_urls) > 0
+            self.markdown.zulip_message.has_image = False  # This is updated in self.add_a
+            # Populate potential_attachment_urls. do_claim_attachments() handles setting has_attachments.
+            self.markdown.zulip_message.potential_attachment_urls = []
+            for url in unique_urls:
+                self.maybe_append_attachment_url(url)
```

Comment:
What's your thinking behind filtering out links inside blockquotes here?  I think it's correct for a quoted link to an attachment to still be claimed...

Reply:
Ahh, I see, it's just a bad variable name; I'll make a new `unique_previewable_urls`.